File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def match(self, a, b):
        return a == '(' and b == ')' or a == '{' and b == '}' or a == '[' and b == ']'

    def isValid(self, s: str) -> bool:
        st = []
        for x in s:
            if x in {'(', '{', '['}:
                st.append(x)
            else:
                if len(st) == 0 or self.match(st[-1], x) == False:
                    return False
                else:
                    st.pop()

        return len(st) == 0
        

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean match(char a, char b){
        return (a == '(' && b == ')') || (a == '{' && b == '}') || (a == '[' && b == ']');
    }
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        for(int i=0; i<s.length(); i++){
            char ch = s.charAt(i);
            if(ch == '(' || ch == '{' || ch == '['){
                st.push(ch);
            }
            else{
                if(st.size() == 0 || (!match(st.peek(), ch))){
                    return false;
                }
                st.pop();
            }
        }
        return st.isEmpty();
    }
}